packages/contracts/contracts/optimistic-ethereum/ovm/StateManagerGasSanitizer.sol

```diff
@@ -0,0 +1,328 @@
+ *         It serves as a proxy between an EM and SM implementation, consuming a fixed amount of gas based on the UPPER_BOUND constants.
+ *
+ *         This allows for OVM gas metering to be independent of the actual consumption of the SM, so that different SM implementations do not change OVM behavior.
+ */
+
+ // TODO: inerit IStateManager after visibility changes
+ // TODO: rename.  Gas sanitizer?
+ // TODO: parammeterize
+contract StateManagerGasSanitizer is ContractResolver {
+
+    /*
+     * Virtual (i.e. Charged by OVM) Gas Cost Constants
+     */
+
+    // Storage
+    uint constant GET_STORAGE_VIRTUAL_GAS_COST = 5000;
+    uint constant SET_STORAGE_VIRTUAL_GAS_COST = 20000;
+    // Nonces
+    uint constant GET_CONTRACT_NONCE_VIRTUAL_GAS_COST = 10000;
+    uint constant SET_CONTRACT_NONCE_VIRTUAL_GAS_COST = 30000;
+    uint constant INCREMENT_CONTRACT_NONCE_VIRTUAL_GAS_COST = 35000;
+    // Code
+    uint constant ASSOCIATE_CODE_CONTRACT_VIRTUAL_GAS_COST = 1000;
+    uint constant REGISTER_CREATED_CONTRACT_VIRTUAL_GAS_COST = 1000;
+    uint constant GET_CODE_CONTRACT_ADDRESS_VIRTUAL_GAS_COST = 1000;
+    uint constant GET_CODE_CONTRACT_HASH_VIRTUAL_GAS_COST = 1000;
+    // Code copy retrieval, linear in code size
+    uint constant GET_CODE_CONTRACT_BYTECODE_VIRTUAL_GAS_COST_PER_BYTE = 10;
+
+    /*
+     * Constant/Upper-bounded Fixed Gas Cost Constants
+     */
+
+    // Storage
+    uint constant GET_STORAGE_GAS_COST_UPPER_BOUND = 200000;
+    uint constant SET_STORAGE_GAS_COST_UPPER_BOUND = 200000;
+    // Nonces
+    uint constant GET_CONTRACT_NONCE_GAS_COST_UPPER_BOUND = 200000;
+    uint constant SET_CONTRACT_NONCE_GAS_COST_UPPER_BOUND = 200000;
+    uint constant INCREMENT_CONTRACT_NONCE_GAS_COST_UPPER_BOUND = 200000;
+    // Code
+    uint constant ASSOCIATE_CODE_CONTRACT_GAS_COST_UPPER_BOUND = 200000;
+    uint constant REGISTER_CREATED_CONTRACT_GAS_COST_UPPER_BOUND = 200000;
+    uint constant GET_CODE_CONTRACT_ADDRESS_GAS_COST_UPPER_BOUND = 200000;
+    uint constant GET_CODE_CONTRACT_HASH_GAS_COST_UPPER_BOUND = 200000;
+    // Code copy retrieval, linear in code size
+    uint constant GET_CODE_CONTRACT_BYTECODE_GAS_COST_UPPER_BOUND = 200000;
+
+
+    /*
+     * Contract Variables
+     */
+
+    uint OVMRefund;
+
+    /*
+     * Modifiers
+     */
+
+
+    /*
+     * Constructor
+     */
+
+    /**
+     * @param _addressResolver Address of the AddressResolver contract.
+     */
+    constructor(
+        address _addressResolver
+    )
+        public
+        ContractResolver(_addressResolver)
+    {}
+
+    /*
+     * Gas Virtualization and Storage
+     */
+
+    // External Initialization and Retrieval Logic
+    function resetOVMRefund() external {
+        OVMRefund = 0;
+    }
+
+    function getOVMRefund() external view returns(uint) {
+        return OVMRefund;
+    }
+
+    // Internal Logic
+    function addToOVMRefund(uint _refund) internal {
+        OVMRefund += _refund;
```

Comment:
Do we need to `require(MAX_UINT - OVMRefund >= _refund, "overflow");`? Based on context, we may not care.

Reply:
I don't think it's possible for any reasonable value for the constants at the top of the file, simply due to the EVM's gas limit.